Room state

fancycade opened the pull request:
We should merge #13 before this.

This PR addresses #7 and #11.

niklaslong commented:
Looks like #13 introduced some conflicts, @fancycade could you rebase?

fancycade commented:
Oops didnt realize the log would be that screwed up. Good thing this is an experiment :rofl:

niklaslong commented on lib/matrix_client.ex:
```diff
@@ -91,10 +91,22 @@ defmodule MatrixClient do
     )
   end
 
+  defp joined_rooms_helper(session, room_ids) do
```
How about something more descriptive, perhaps `map_joined_rooms` or `list_joined_rooms`?

niklaslong commented on lib/matrix_client.ex:
```diff
@@ -146,11 +166,51 @@ defmodule MatrixClient do
+    {:ok, token} = Session.get(pid, :token)
+    aliases = Session.get_aliases(pid)
+
+    new_id =
```
As an aside, I'm seeing the following block a lot—refactor opportunity in future.

niklaslong commented on lib/matrix_client/session.ex:
```diff
@@ -5,7 +5,9 @@ defmodule MatrixClient.Session do
   def start_link(url) do
-    Agent.start_link(fn -> %{url: url, rooms: %{}, invites: %{}} end)
+    Agent.start_link(fn ->
+      %{url: url, rooms: %{}, invites: %{}, aliases: %{ids: %{}, aliases: %{}}}
```
What are the `ids` in the `aliases` map?

niklaslong commented on lib/matrix_client/session.ex:
```diff
@@ -120,8 +154,40 @@ defmodule MatrixClient.Session do
+    end
+  end
+
+  defp alias_helper_ids(aliases, room_id, a) do
```
Not sure what these functions are doing just by reading the name? In general I'm not a big fan of using `helper` as a descriptor as it's so vague. Thoughts?

niklaslong reviewed (changes requested):
Great work—I've left a few comments. Once this is merged I'll review the project and its status in relation to the SDK.

fancycade commented:
@niklaslong Updated based on feedback.

Changed any helper names or refactored.

Moved ids into its own map. The reason for it  is that internally there are times to translate an alias to an id, and then the alias map is to translate an id to an alias (typically for external reasons like listing joined rooms). Can't decide if the naming is backwards or not :laughing:
